### PR TITLE
feat(source): add metrics for throughput of each partition

### LINKED ISSUE #3504: metric for input rate

### PR DESCRIPTION
**Pull request by tabVersion**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

***PLEASE DO NOT LEAVE THIS EMPTY !!!***

Please explain **IN DETAIL** what the changes are in this PR and why they are needed:

as title, the metric records throughput before data sending to channel in connector source, which indicates the max throughput of source. 

- Summarize your change (**mandatory**)
- How does this PR work? Need a brief introduction for the changed logic (optional)
- Describe clearly one logical change and avoid lazy messages (optional)
- Describe any limitations of the current code (optional)
- Add the 'user-facing changes' label if your PR contains changes that are visible to users (optional)

## Checklist

- [x] I have written necessary docs and comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Refer to a related PR or issue link (optional)

close #3504

### CODE REVIEW

File: src/stream/src/executor/source.rs [hunk 1]
```diff
@@ -276,7 +277,12 @@ impl<S: StateStore> SourceExecutor<S> {
                 .await
                 .map(SourceStreamReaderImpl::TableV2),
             SourceImpl::Connector(c) => c
-                .stream_reader(state, self.column_ids.clone())
+                .stream_reader(
+                    state,
+                    self.column_ids.clone(),
+                    self.metrics.registry.clone(),
```
Comment (tabVersion) [inline src/stream/src/executor/source.rs:283]: @wcy-fdu I had some trouble with building `SourceMetrics` by this registry. pls take a look

**Comment by skyzh** [comment]
Is this metrics available in redpanda / kafka-exporter? I wonder if it is necessary also to monitor on our side… Rest LGTM

File: src/source/src/connector_source.rs [hunk 2]
```diff
@@ -113,21 +144,31 @@ impl InnerConnectorSourceReader {
...
+                }
+                Ok(None) => {
+                    log::warn!("connector reader {} stream stopped", id);
+                    break;
+                }
+                Ok(Some(msg)) => {
+                    self.metrics
+                        .partition_input_count
+                        .with_label_values(&[
+                            self.context.actor_id.to_string().as_str(),
```
Comment (Sunt-ing) [inline src/source/src/connector_source.rs:166]: To reduce overhead, a new field "actor_id_str" can be added to self.context and then used here. In this way, frequent type conversions can be avoided.

File: src/source/src/connector_source.rs [hunk 3]
```diff
@@ -113,21 +144,31 @@ impl InnerConnectorSourceReader {
...
+                            self.context.source_id.to_string().as_str(),
```
Comment (Sunt-ing) [inline src/source/src/connector_source.rs:167]: and used here

File: src/source/src/connector_source.rs [hunk 4]
```diff
@@ -113,21 +144,31 @@ impl InnerConnectorSourceReader {
...
+                            id.as_str(),
```
Comment (Sunt-ing) [inline src/source/src/connector_source.rs:168]: similar change can be done here

**Review by Sunt-ing — APPROVED** [review]
Generally LGTM!

**Comment by Sunt-ing** [comment]
It's normal for different sources may have different metrics. Take Flink connector metrics as an example: https://nightlies.apache.org/flink/flink-docs-master/docs/ops/metrics/#connectors

**Comment by tabVersion** [comment]
> Is this metrics available in redpanda / kafka-exporter? I wonder if it is necessary also to monitor on our side

Sure kafka provides metrics like this but currently, we do not commit to kafka so it does not know the consumption process. cc @shanicky

**Review by KeXiangWang — APPROVED** [review]
LGTM

**Pull request merged**